pkg/merkle/proof.go

```diff
@@ -0,0 +1,345 @@
+
+				if nextIsSibling {
+					// Get left sibling from queue
+					leftHash = hashQueue[readIndex]
+					readIndex = (readIndex + 1) % numElements
+				} else {
+					// Get left sibling from proof
+					leftHash, err = p.getNextProof(&proofIndex)
+					if err != nil {
+						return nil, err
+					}
+				}
+			} else {
+				// Current node is left child
+				leftHash = hashQueue[readIndex]
+				readIndex = (readIndex + 1) % numElements
+
+				// Always get right sibling from proof
+				rightHash, err = p.getNextProof(&proofIndex)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			if leftHash == nil || rightHash == nil {
+				return nil, ErrNilProof
+			}
+
+			// Compute parent hash and add to queue
+			parentIndex := currentIndex >> 1
+			parentHash := HashNode(leftHash, rightHash)
+			nodeQueue[writeIndex] = parentIndex
+			hashQueue[writeIndex] = parentHash
+			writeIndex = (writeIndex + 1) % numElements
+		}
+
+		// Update level tracking when processing the lowest index.
+		if currentIndex == lowestTreeIndex || nextNodeIndex == lowestTreeIndex {
+			lowestTreeIndex >>= 1
+			upperBound >>= 1
+		}
+	}
+}
+
+// getNextProof safely retrieves the next proof and increments the index.
+func (p *MultiProof) getNextProof(proofIndex *int) ([]byte, error) {
+	if *proofIndex >= len(p.proofs) {
+		return nil, ErrNilProof
+	}
+	proof := p.proofs[*proofIndex]
+	*proofIndex++
+	return proof, nil
+}
+
+// validate performs common validation for Merkle proofs.
+func (p *MultiProof) validate() error {
+	if len(p.elements) == 0 {
+		return ErrNoElements
+	}
+
+	if p.leafCount <= 0 {
+		return ErrInvalidLeafCount
+	}
+
+	if err := validateIndices(p.elements.Indices(), p.leafCount); err != nil {
+		return err
+	}
+
+	for _, elem := range p.elements {
+		if elem.value == nil {
+			return ErrNilElement
+		}
+	}
+
+	for _, proof := range p.proofs {
+		if proof == nil {
+			return ErrNilProof
+		}
+	}
+
+	isPartialProof := len(p.elements) < p.leafCount
+	isNonTrivialTree := p.leafCount > 1
+	needsProofs := isPartialProof && isNonTrivialTree
+
+	if needsProofs && len(p.proofs) == 0 {
+		return ErrNoProofs
+	}
+
+	return nil
+}
+
+// makeIndexedValues creates indexed values from elements and their indices.
+func makeIndexedValues(elements [][]byte, indices []int) indexedValues {
+	result := make(indexedValues, len(indices))
+	for i, idx := range indices {
+		result[i] = indexedValue{
+			value: elements[idx],
+			index: idx,
+		}
+	}
+	return result
+}
+
+// makeIndices returns a slice of ascending ordered indices for the given starting index and count.
+func makeIndices(startingIndex, count int) ([]int, error) {
+	if startingIndex < 0 || count <= 0 {
+		return nil, ErrInvalidRange
+	}
+
+	indices := make([]int, count)
+	for i := 0; i < count; i++ {
+		indices[i] = startingIndex + i
+	}
+
+	return indices, nil
+}
+
+// validateIndices validates the indices slice of a proof.
+func validateIndices(indices []int, leafCount int) error {
```

Comment:
What about gaps in the list. For example, if I provide `1, 2, 4`. Those would not have duplicates, or out of bounds, but should be invalid for a sequential multi-proof.

Or is the goal to handle that validation elsewhere so that we can generate adversarial test cases?

Reply:
Currently we support two multi proofs; sequential and indices based. That's why we don't explicitly check the provided indices are always sequential.

We can be more opinionated and just do sequential indices.

Reply:
For sequential proofs, we need to check _somewhere_ that they are actually sequential or someone could exclude a Payer from settlement. Unclear whether to me whether that should be here or elsewhere.